beacon-chain/db/kv/slasher_test.go

```diff
@@ -482,6 +483,139 @@ func TestStore_PruneAttestations(t *testing.T) {
+				createAttestationWrapper(5, 6, []uint64{5}, []byte{1}),
+			},
+			indices: []types.ValidatorIndex{2, 3, 4, 5},
+			expected: []*slashpb.HighestAttestation{
+				{
+					ValidatorIndex:     2,
+					HighestSourceEpoch: 0,
+					HighestTargetEpoch: 3,
+				},
+				{
+					ValidatorIndex:     3,
+					HighestSourceEpoch: 1,
+					HighestTargetEpoch: 4,
+				},
+				{
+					ValidatorIndex:     4,
+					HighestSourceEpoch: 2,
+					HighestTargetEpoch: 3,
+				},
+				{
+					ValidatorIndex:     5,
+					HighestSourceEpoch: 5,
+					HighestTargetEpoch: 6,
+				},
+			},
+		},
+		{
+			name: "should get correct highest att for multiple shared atts with diff histories",
+			attestationsInDB: []*slashertypes.IndexedAttestationWrapper{
+				createAttestationWrapper(1, 4, []uint64{2, 3}, []byte{1}),
+				createAttestationWrapper(2, 5, []uint64{3, 5}, []byte{1}),
+				createAttestationWrapper(4, 5, []uint64{1, 2}, []byte{1}),
+				createAttestationWrapper(6, 7, []uint64{5}, []byte{1}),
+			},
+			indices: []types.ValidatorIndex{2, 3, 4, 5},
+			expected: []*slashpb.HighestAttestation{
+				{
+					ValidatorIndex:     2,
+					HighestSourceEpoch: 4,
+					HighestTargetEpoch: 5,
+				},
+				{
+					ValidatorIndex:     3,
+					HighestSourceEpoch: 2,
+					HighestTargetEpoch: 5,
+				},
+				{
+					ValidatorIndex:     5,
+					HighestSourceEpoch: 6,
+					HighestTargetEpoch: 7,
+				},
+			},
+		},
+	}
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			beaconDB := setupDB(t)
+			require.NoError(t, beaconDB.SaveAttestationRecordsForValidators(ctx, tt.attestationsInDB))
+
+			highestAttestations, err := beaconDB.HighestAttestations(ctx, tt.indices)
+			require.NoError(t, err)
+			require.Equal(t, len(tt.expected), len(highestAttestations))
+			for i, existing := range highestAttestations {
+				require.DeepEqual(t, existing, tt.expected[i])
+			}
+		})
+	}
+}
+
+func BenchmarkHighestAttestations(b *testing.B) {
+	count := 10000
+	valsPerAtt := 100
+	indicesPerAtt := make([][]uint64, count)
+	for i := 0; i < count; i++ {
+		indicesForAtt := make([]uint64, valsPerAtt)
+		for r := i * count; r < valsPerAtt*(i+1); r++ {
+			indicesForAtt[i] = uint64(r)
+		}
+		indicesPerAtt[i] = indicesForAtt
+	}
+	atts := make([]*slashertypes.IndexedAttestationWrapper, count)
+	for i := 0; i < count; i++ {
+		atts[i] = createAttestationWrapper(types.Epoch(i), types.Epoch(i+2), indicesPerAtt[i], []byte{})
+	}
+
+	ctx := context.Background()
+	beaconDB := setupDB(b)
+	require.NoError(b, beaconDB.SaveAttestationRecordsForValidators(ctx, atts))
+
+	allIndices := make([]types.ValidatorIndex, valsPerAtt*count)
+	for i := 0; i < count; i++ {
+		indicesForAtt := make([]types.ValidatorIndex, valsPerAtt)
+		for r := 0; r < valsPerAtt; r++ {
+			indicesForAtt[r] = types.ValidatorIndex(atts[i].IndexedAttestation.AttestingIndices[r])
+		}
+		allIndices = append(allIndices, indicesForAtt...)
+	}
+	b.ReportAllocs()
```

Comment:
We need to only bench starting from here. Do b.StopTimer() at the top of the func Benchmark and b.StartTimer right before this for loop